org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java

```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
 
     assertEqualsWithComparisonFailure(expected, actual);
   }
+
+  /**
+   * Tests applied value migrations provided in context.
+   */
+  @Test
+  public void testAppliedValueMigrationHandlers() {
+    // raw DoEntity data object
+    IDoEntity original = BEANS.get(DoEntityBuilder.class)
+        .put("_type", "charlieFixture.RoomFixture")
+        .put("_typeVersion", CharlieFixture_5.VERSION.unwrap()) // latest version, no structure migration will be applied
+        .put("name", "example")
+        .put("areaInSquareMeter", 10)
+        .build();
+
+    // no value migrations will match the provided data object
+    // RoomSizeValueMigrationHandler is ignored by default in s_migrationContext
+    DoStructureMigratorResult<RoomFixtureDo> result = s_migrator.migrateDataObject(s_migrationContext, original, RoomFixtureDo.class);
+
+    assertFalse(result.isChanged()); // no changes (structure nor values changed)
+
+    IDoEntity expected = BEANS.get(RoomFixtureDo.class)
+        .withName("example")
+        .withAreaInSquareMeter(10);
+
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+  }
+
+  /**
+   * Tests non-idempotent value migration with {@link RoomSizeValueMigrationHandler}.<br>
+   * Migration handler matches and replaces whole data object (RoomFixtureDo)
+   */
+  @Test
+  public void testNonIdempotentValueMigration() {
+    // raw DoEntity data object
+    IDoEntity original = BEANS.get(DoEntityBuilder.class)
+        .put("_type", "charlieFixture.RoomFixture")
+        .put("_typeVersion", CharlieFixture_5.VERSION.unwrap()) // latest version, no structure migration will be applied
+        .put("name", "example")
+        .put("areaInSquareMeter", 10)
+        .build();
+
+    DoStructureMigrationContext context = BEANS.get(DoStructureMigrationContext.class)
+        .putGlobal(BEANS.get(DoValueMigrationIdsContextData.class)
+            .withAppliedValueMigrationIds(Collections.emptySet())); // run all value migrations
+
+    // migration run 1
+    DoStructureMigratorResult<RoomFixtureDo> result = s_migrator.migrateDataObject(context, original, RoomFixtureDo.class);
+
+    assertTrue(result.isChanged());
+
+    IDoEntity expected = BEANS.get(RoomFixtureDo.class)
+        .withName("example")
+        .withAreaInSquareMeter(20); // increased by 10 through RoomSizeValueMigrationHandler
+
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+
+    // migration run 2: data object is changed again, due to non-idempotent value migration
+    result = s_migrator.migrateDataObject(context, result.getDataObject(), RoomFixtureDo.class);
```

Comment:
called with already typed object here, so no structure migrations will be applied, but should be okay for test